### PR TITLE
Improve performance of MLflow logging

### PR DESCRIPTION
**Pull request by dbczumar**
# What does this PR do?

Improve performance of the Composer MLflow logging integration by leveraging a queuing, async client

**TESTING**: Performed manual QA on a Databricks workspace with LLM-foundry and confirmed that all expected run data is logged

Screenshot 2023-08-22 at 2 48 42 PM
Screenshot 2023-08-22 at 2 48 49 PM

# What issue(s) does this change relate to?

# Before submitting
- [X] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [X] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [X] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Review by eracah — APPROVED** [review]
LGTM! Will defer to @dakinggg or @mvpatel2000 for final lookover

**Comment by eracah** [comment]
@dbczumar, plz wait to merge!

File: composer/loggers/mlflow_logger.py
Comment (mvpatel2000) [inline composer/loggers/mlflow_logger.py:80]:
```suggestion
        elif exp := self._mlflow_client.get_experiment_by_name(name=self.experiment_name):
```
nit: can we not use walrus? 

we officially only support >=3.8, but it would be nice to be backwards compatible (up until now we are!) for community users who are using lower python versions for compatibility with other libraries

Comment (mvpatel2000) [inline composer/loggers/mlflow_logger.py:80]: Should this all be gated with `self._enabled`? We probably only want to set up if we're logging from this rank, right?

**Review by mvpatel2000 — COMMENTED** [review]
A few things I'm not quite sure about before we merge :)

File: composer/loggers/mlflow_logger.py
Comment (dbczumar) [reply composer/loggers/mlflow_logger.py]: Absolutely - I've gone ahead and added that gating :)

Comment (dbczumar) [reply composer/loggers/mlflow_logger.py]: Great point! I've removed the walrus! I am no longer the walrus. 🦭

File: composer/loggers/mlflow_logger.py
Comment (mvpatel2000) [inline composer/loggers/mlflow_logger.py:76]:
```suggestion
            # Set experiment. We use MlflowClient for experiment retrieval and creation
```
Nit

File: composer/loggers/mlflow_logger.py
Comment (mvpatel2000) [inline composer/loggers/mlflow_logger.py:87]:
```suggestion
```
Fine to not have since it's last line of function

File: composer/loggers/mlflow_logger.py
Comment (mvpatel2000) [inline composer/loggers/mlflow_logger.py:62]: Should this be gated behind the enable check 3 lines down as well? Not sure if this pings server or something 🤷

File: composer/loggers/mlflow_logger.py
Comment (mvpatel2000) [inline composer/loggers/mlflow_logger.py:100]:
```suggestion
        # Start run
```
nit (which is really a nit from the previous code :P)

File: composer/loggers/mlflow_logger.py
Comment (mvpatel2000) [inline composer/loggers/mlflow_logger.py:136]:
```suggestion
            # We use MlflowClient for run termination because MlflowAutologgingQueueingClient's
```
nit

**Review by mvpatel2000 — APPROVED** [review]
LGTM! A few minor comments -- apologies, I should have been more thorough on the first pass.

Also, can you please manually test this at least once before merging?

File: composer/loggers/mlflow_logger.py
Comment (dbczumar) [reply composer/loggers/mlflow_logger.py]: Fixed! :)

File: composer/loggers/mlflow_logger.py
Comment (dbczumar) [reply composer/loggers/mlflow_logger.py]: Fixed! :)

**Review by dbczumar — COMMENTED** [review]
@mvpatel2000 Thanks for the feedback and approval! I took care of those comments and will run manual QA once more before merge

**Pull request merged**
